### DISCUSSION TITLE
KeyError: 'DISKANN'

### DISCUSSION THREAD

**Discussion by LMERose** [milvus-io/milvus#33346] [Q&A and General discussion]
hello, I'm trying to change my index from HNSW to DISKNN, but it looks like wrong.I want to know if it's because langchain doesn't support disknn?
  File "/home/knowadmin/anaconda3/envs/LLM-test/lib/python3.8/site-packages/langchain_community/vectorstores/milvus.py", line 454, in _create_search_params
    self.search_params = self.default_search_params[index_type]
KeyError: 'DISKANN'

**Comment by yhmo** [comment]
https://milvus.io/docs/disk_index.md#Prerequisites

To enable DISKANN, make sure the queryNode.enableDisk is true in the milvus.yaml.

**Reply by LMERose** [reply]
I have modified this configuration, but errors still occur.

# Related configuration of queryNode, used to run hybrid search between vector and scalar data.
queryNode:
  dataSync:
    flowGraph:
      maxQueueLength: 16 # Maximum length of task queue in flowgraph
      maxParallelism: 1024 # Maximum number of tasks executed in parallel in the flowgraph
  stats:
    publishInterval: 1000 # Interval for querynode to report node information (milliseconds)
  segcore:
    cgoPoolSizeRatio: 2.0 # cgo pool size ratio to max read concurrency
    knowhereThreadPoolNumRatio: 4
    # Use more threads to make better use of SSD throughput in disk index.
    # This parameter is only useful when enable-disk = true.
    # And this value should be a number greater than 1 and less than 32.
    chunkRows: 128 # The number of vectors in a chunk.
    exprEvalBatchSize: 8192 # The batch size for executor get next
    interimIndex: # build a vector temperate index for growing segment or binlog to accelerate search
      enableIndex: true
      nlist: 128 # segment index nlist
      nprobe: 16 # nprobe to search segment, based on your accuracy requirement, must smaller than nlist
      memExpansionRate: 1.15 # the ratio of building interim index memory usage to raw data
      buildParallelRate: 0.5 # the ratio of building interim index parallel matched with cpu num
  loadMemoryUsageFactor: 1 # The multiply factor of calculating the memory usage while loading segments
  enableDisk: true # enable querynode load disk index, and search on disk index
  maxDiskUsagePercentage: 95
  cache:
    enabled: true # deprecated, TODO: remove it
    memoryLimit: 2147483648 # 2 GB, 2 * 1024 *1024 *1024 # deprecated, TODO: remove it
    readAheadPolicy: willneed # The read ahead policy of chunk cache, options: `normal, random, sequential, willneed, dontneed`

**Comment by yhmo** [comment]
Which version of your milvus?
Which version of the langchain?
Which version of the pymilvus?

**Reply by LMERose** [reply]
pymilvus---2.4.0
langchain---0.1.0
langchain-community---0.0.20
milvus-standalone-docker-compose-gpu---3.5

**Comment by yhmo** [comment]
The error is caused by the default_search_params, default search params for DISKANN is missed: https://github.com/langchain-ai/langchain/blob/2edb5122825591bf86791055305630816247b5bb/libs/community/langchain_community/vectorstores/milvus.py#L148

Specify an empty search_prams then it works:

```
from langchain_community.vectorstores import Milvus
from tests.integration_tests.vectorstores.fake_embeddings import (
    FakeEmbeddings,
    fake_texts,
)


if __name__ == '__main__':
    store = Milvus.from_texts(
        texts=fake_texts,
        embedding=FakeEmbeddings(),
        metadatas=None,
        ids=None,
        connection_args={"host": "127.0.0.1", "port": "19530"},
        collection_name="langchain_test",
        drop_old=True,
        index_params={
            "metric_type": "L2",
            "index_type": "DISKANN",
            "params": {},
        },
        search_params={},
    )
    res = store.similarity_search("bar", k=2)
    print(res)
```

**Reply by LMERose** [reply]
tanks very much! it is work now! I have another question, how can I check the disk space and memory space usage of each collection?

**Reply by yhmo** [reply]
If your milvus is a cluster, you can deploy a monitoring system to observe disk/memory usage. But it only provides the entire disk/memory usage, not for distinct collection.
If your milvus is a standalone, there is no interface to check disk/memory usage of each collection.


It is complicated to say "how much disk space of a collection".

1. Note that each modern database has a WAL(write-ahead-log) component. All the new data is firstly stored into the WAL for data safety. If the database crashes, it can restore the lost data from the WAL.
Milvus also has a WAL component. For Milvus Cluster, Pulsar is the WAL component. For Milvus Standalone, a local path is used to store WAL.

After you start a stadalone by docker-compose, you can see there is a folder named "volumes" created beside the docker-compose.yaml.
Browse the directory tree you will see:
➜  volumes tree -L 2
.
├── etcd
│   └── member
├── milvus
│   ├── data
│   ├── rdb_data
│   └── rdb_data_meta_kv
└── minio
    └── a-bucket

volumes/milvus/rdb_data is the path to store WAL, managed by an embedded RocksDB. If you continually insert data, you will see the size of this directory continually increase. RocksDB compresses the WAL and has its compaction behavior. 
The retention is controlled by the configurations in the milvus.yaml:
```
rocksmq:
  retentionTimeInMinutes: 4320 # 3 days, 3 * 24 * 60 minutes, The retention time of the message in rocksmq.
```
Default retention time is 3 days.



2. Milvus delete action is "soft-delete". If a collection(or a partition, or a segment) is deleted, it only marks it as "to-be-deleted", the data files are still in storage, waiting for the GC process.
Milvus GC process configurations in milvus.yaml:
```
dataCoord:
  gc:
    interval: 3600 # gc interval in seconds
    missingTolerance: 86400 # file meta missing tolerance duration in seconds, default to 24hr(1d)
    dropTolerance: 10800 # file belongs to dropped entity tolerance duration in seconds. 3600
```

GC process detects the soft-delete items for each 1 hour(gc.interval). Some items might be kept in storage for gc.dropTolerance. So, a soft-delete file might be kept in disk for 2~4 hours. 
To reduce disk usage, you can:
change the rocksmq.retentionTimeInMinutes to 10  (I recommend this value to be larger than 10. Small value might lead to data lost when you have intensive insertion and milvus crashes)
change the dataCoord.gc.dropTolerance to 1800 (I recommend this value to be larger than 1800)


3. If index is built, milvus requires extra disk space to store index. Milvus builds an independent index for each segment. If a segment is indexed and collection is loaded, the segment's index will be loaded into memory. If a segment is not indexed and the collection is loaded, the segment's original data will be loaded into memory.
Typically, the index size compare to the original vectors size:
- IVF_FLAT/HNSW, a bit larger than the original vectors size
- IVF_SQ8/IVF_PQ, 25% ~ 30% of the original vectors size, the recall rate of IVF_PQ is not good.
- DISKANN, 1/4 ~ 1/6 of the original vectors size, performance is not good, requires high-performance hard disk.
